File: validate.js
const fs = require('fs');
const path = require('path');
const eng = require('./desktop/src/utils/edergawiEngine');

const GREG_NUM = {January:1,February:2,March:3,April:4,May:5,June:6,July:7,August:8,September:9,October:10,November:11,December:12};

const csv = fs.readFileSync(path.join(__dirname, 'verify.csv'), 'utf-8');
const lines = csv.split('\n').map(l => l.trim()).filter(l => l && !l.startsWith('Arabic Name'));

let currentYear = null;
const expected = [];

for (const line of lines) {
  const yearMatch = line.match(/^(\d{4}):$/);
  if (yearMatch) {
    currentYear = parseInt(yearMatch[1]);
    continue;
  }
  if (currentYear === null) continue;
  const parts = line.split(',');
  if (parts.length !== 4) continue;
  const [arabic, english, gregMonth, dayStr] = parts;
  const day = parseInt(dayStr);
  const gregM = GREG_NUM[gregMonth];
  if (!gregM) {
    console.log(`WARN: Unknown Gregorian month "${gregMonth}" for ${english} ${currentYear}`);
    continue;
  }
  expected.push({ year: currentYear, english, arabic, gregMonth: gregM, day });
}

let totalMatch = 0, totalMismatch = 0, totalExpected = 0;
const mismatches = [];

const years = [...new Set(expected.map(e => e.year))].sort((a,b) => a-b);

for (const year of years) {
  const yearExpected = expected.filter(e => e.year === year);
  const aeYear = eng.gregToAEYear(year);

  process.stdout.write(`\n${year} (AE ${aeYear}): `);

  let yearTotal = 0, yearMatch = 0, yearMiss = 0;

  for (const exp of yearExpected) {
    // Determine the engine's month index for this expected month
    // Try looking up by English name in this Gregorian year
    let found = false;

    const months = eng.getMonthsForYear(year);
    for (let mi = 0; mi < months.length; mi++) {
      const m = months[mi];
      const mName = m.english.replace('Alshahr', 'Alshar');
      const eName = exp.english.replace('Alshahr', 'Alshar');
      if (mName === eName) {
        found = true;
        totalExpected++;
        yearTotal++;

        // The expected date from CSV
        const expDate = new Date(year, exp.gregMonth - 1, exp.day);
        // The engine's start date for this month
        const engDate = m.startDate;

        const expStr = `${String(exp.gregMonth).padStart(2,'0')}-${String(exp.day).padStart(2,'0')}`;
        const engStr = `${String(engDate.getMonth()+1).padStart(2,'0')}-${String(engDate.getDate()).padStart(2,'0')}`;
        const diff = Math.round((engDate - expDate) / 86400000);

        if (diff === 0) {
          totalMatch++;
          yearMatch++;
        } else {
          totalMismatch++;
          yearMiss++;
          mismatches.push({ year, month: exp.english, expDate: expStr, engDate: engStr, diff });
        }
        break;
      }
    }
    if (!found) {
      process.stdout.write(`[${exp.english} NOT FOUND] `);
    }
  }

  const pct = yearTotal > 0 ? Math.round(yearMatch / yearTotal * 100) : 0;
  process.stdout.write(`${yearMatch}/${yearTotal} (${pct}%)`);
}

const total = totalMatch + totalMismatch;
const pct = total > 0 ? Math.round(totalMatch / total * 100) : 0;
console.log(`\n\n=== SUMMARY ===`);
console.log(`Total: ${totalMatch}/${total} matched (${pct}%)`);
console.log(`Mismatches: ${totalMismatch}`);

if (mismatches.length > 0) {
  console.log(`\nDetailed mismatches (first 30):`);
  for (const m of mismatches.slice(0, 30)) {
    console.log(`  ${m.year} ${m.month}: expected ${m.expDate}, got ${m.engDate} (diff ${m.diff}d)`);
  }
  if (mismatches.length > 30) {
    console.log(`  ... and ${mismatches.length - 30} more`);
  }
}

File: validate.py
from edergawi_calendar_utils import get_edergawi_year

GREG_REV = {
    'January': 1, 'February': 2, 'March': 3, 'April': 4,
    'May': 5, 'June': 6, 'July': 7, 'August': 8,
    'September': 9, 'October': 10, 'November': 11, 'December': 12
}
GREG_NAMES = {v: k for k, v in GREG_REV.items()}
BASE_YEAR = 2022

def greg_to_ae(gy):
    if gy >= BASE_YEAR:
        return gy - BASE_YEAR + 1
    return -(BASE_YEAR - gy)

lines = open('verify.csv').read().strip().split('\n')
lines = [l for l in lines if not l.startswith('Arabic Name')]

errors = []
total = 0
current_year = None

for line in lines:
    line = line.strip()
    if not line:
        continue
    if line.endswith(':'):
        current_year = int(line.rstrip(':'))
        continue
    parts = line.split(',')
    if len(parts) < 4:
        continue
    arabic = parts[0].strip()
    english = parts[1].strip()
    greg_month_str = parts[2].strip()
    day_str = parts[3].strip()
    day = int(day_str)
    greg_month_num = GREG_REV[greg_month_str]
    total += 1
    ae = greg_to_ae(current_year)
    yi = get_edergawi_year(ae)
    found = None
    for m in yi['months']:
        if m['english'] == english:
            found = m
            break
    if found is None:
        errors.append(f'{current_year} {english}: month not found')
        continue
    actual_gy = found['gregYear']
    actual_gm_str = found['gregorian']
    actual_day = found['startDay']
    actual_gm = GREG_REV[actual_gm_str]
    expected_gy = current_year
    if actual_gy != expected_gy:
        errors.append(f'{current_year} {english}: expected year {expected_gy} got {actual_gy}')
        continue
    if actual_gm != greg_month_num:
        errors.append(f'{current_year} {english}: expected greg month {greg_month_str} got {actual_gm_str}')
        continue
    if actual_day != day:
        diff = actual_day - day
        errors.append(f'{current_year} {english}: expected day {day} got {actual_day} diff={diff}')
        continue

match_count = total - len(errors)
pct = 100.0 * match_count / total if total > 0 else 0
print(f'Total: {total}, Matches: {match_count}, Errors: {len(errors)}, Match rate: {pct:.1f}%')
if errors:
    print('\nErrors:')
    for e in errors:
        print(e)
